File: typescript/src/autonomousFinancial.ts
/**
 * Autonomous Financial Execution — governance for AI-driven financial actions.
 *
 * Defines bounded authority, execution ceilings, and runtime verification
 * requirements for autonomous agents performing financial operations:
 * refunds, procurement, cloud-cost optimization, vendor payments, etc.
 *
 * Wire-stable as `autonomous_financial.v1`.
 */

import type { CurrencyCode, FinancialActionType, FinancialRiskTier } from "./financialAction.js";

/** Authority bounds for an autonomous financial agent. */
export interface AutonomousExecutionBounds {
  readonly bounds_id: string;
  readonly org_id: string;
  readonly agent_id: string;
  readonly agent_name: string;
  /** Action types this agent is permitted to execute autonomously. */
  readonly permitted_action_types: readonly FinancialActionType[];
  readonly ceilings: readonly ExecutionCeiling[];
  readonly daily_aggregate_ceiling: number;
  readonly aggregate_currency: CurrencyCode;
  /** Maximum risk tier the agent may autonomously execute. */
  readonly max_risk_tier: FinancialRiskTier;
  readonly require_runtime_verification: boolean;
  readonly anomaly_detection_enabled: boolean;
  readonly created_at: string;
  readonly expires_at: string | null;
  readonly active: boolean;
}

/** Per-action-type execution ceiling. */
export interface ExecutionCeiling {
  readonly action_type: FinancialActionType;
  readonly per_execution_max: number;
  readonly currency: CurrencyCode;
  readonly max_daily_count: number | null;
  readonly require_permit: boolean;
}

/** Record of an autonomous execution attempt. */
export interface AutonomousExecutionRecord {
  readonly record_id: string;
  readonly agent_id: string;
  readonly org_id: string;
  readonly action_type: FinancialActionType;
  readonly action_value: number;
  readonly currency: CurrencyCode;
  readonly permitted: boolean;
  readonly denial_reason: string | null;
  readonly permit_id: string | null;
  readonly anomaly_detected: boolean;
  readonly anomaly_description: string | null;
  readonly attempted_at: string;
  readonly executed_at: string | null;
}

/** Result of checking whether an autonomous execution is within bounds. */
export interface AutonomousExecutionCheckResult {
  readonly permitted: boolean;
  readonly action_type_permitted: boolean;
  readonly within_execution_ceiling: boolean;
  readonly within_daily_aggregate: boolean;
  readonly within_risk_tier: boolean;
  readonly bounds_active: boolean;
  readonly bounds_not_expired: boolean;
  readonly applicable_ceiling: ExecutionCeiling | null;
  readonly denial_reason: string | null;
  readonly violations: readonly string[];
}

const RISK_TIER_ORDER: Record<FinancialRiskTier, number> = {
  low:      1,
  medium:   2,
  high:     3,
  critical: 4,
};

/**
 * Check whether an autonomous execution is within declared bounds.
 */
export function checkAutonomousBounds(params: {
  bounds: AutonomousExecutionBounds;
  actionType: FinancialActionType;
  actionValue: number;
  currency: CurrencyCode;
  riskTier: FinancialRiskTier;
  currentDailyAggregate: number;
  currentDailyCount: Partial<Record<string, number>>;
  now?: Date;
}): AutonomousExecutionCheckResult {
  const violations: string[] = [];
  const nowIso = (params.now ?? new Date()).toISOString();

  const boundsActive = params.bounds.active;
  if (!boundsActive) violations.push("agent execution bounds are inactive");

  const boundsNotExpired =
    params.bounds.expires_at === null || params.bounds.expires_at > nowIso;
  if (!boundsNotExpired) {
    violations.push(`agent bounds expired at ${params.bounds.expires_at}`);
  }

  const actionTypePermitted = (params.bounds.permitted_action_types as string[]).includes(
    params.actionType,
  );
  if (!actionTypePermitted) {
    violations.push(`action type ${params.actionType} not in agent's permitted set`);
  }

  const applicableCeiling =
    params.bounds.ceilings.find((c) => c.action_type === params.actionType) ?? null;

  let withinExecutionCeiling = true;
  if (applicableCeiling !== null) {
    if (params.actionValue > applicableCeiling.per_execution_max) {
      withinExecutionCeiling = false;
      violations.push(
        `value ${params.actionValue} exceeds per-execution ceiling ${applicableCeiling.per_execution_max} ${applicableCeiling.currency}`,
      );
    }
    if (applicableCeiling.max_daily_count !== null) {
      const todayCount = params.currentDailyCount[params.actionType] ?? 0;
      if (todayCount >= applicableCeiling.max_daily_count) {
        withinExecutionCeiling = false;
        violations.push(
          `daily count ${todayCount} at or exceeds limit ${applicableCeiling.max_daily_count} for ${params.actionType}`,
        );
      }
    }
  }

  const withinDailyAggregate =
    params.currentDailyAggregate + params.actionValue <= params.bounds.daily_aggregate_ceiling;
  if (!withinDailyAggregate) {
    violations.push(
      `daily aggregate ${params.currentDailyAggregate + params.actionValue} would exceed ceiling ${params.bounds.daily_aggregate_ceiling} ${params.bounds.aggregate_currency}`,
    );
  }

  const withinRiskTier =
    RISK_TIER_ORDER[params.riskTier] <= RISK_TIER_ORDER[params.bounds.max_risk_tier];
  if (!withinRiskTier) {
    violations.push(
      `action risk tier ${params.riskTier} exceeds agent max ${params.bounds.max_risk_tier}`,
    );
  }

  const permitted =
    boundsActive &&
    boundsNotExpired &&
    actionTypePermitted &&
    withinExecutionCeiling &&
    withinDailyAggregate &&
    withinRiskTier;

  return {
    permitted,
    action_type_permitted: actionTypePermitted,
    within_execution_ceiling: withinExecutionCeiling,
    within_daily_aggregate: withinDailyAggregate,
    within_risk_tier: withinRiskTier,
    bounds_active: boundsActive,
    bounds_not_expired: boundsNotExpired,
    applicable_ceiling: applicableCeiling,
    denial_reason: permitted ? null : (violations[0] ?? "execution out of bounds"),
    violations,
  };
}

/**
 * Detect a potential anomaly in autonomous execution.
 * Returns a description when an anomaly is detected, or null.
 */
export function detectAutonomousAnomaly(params: {
  actionValue: number;
  historicalMeanValue: number;
  historicalStdDev: number;
  recentExecutionCount: number;
  burstThreshold: number;
  isOffHours: boolean;
}): { anomalyDetected: boolean; description: string | null } {
  const zScore =
    params.historicalStdDev > 0
      ? Math.abs(params.actionValue - params.historicalMeanValue) / params.historicalStdDev
      : 0;

  if (zScore > 3) {
    return {
      anomalyDetected: true,
      description: `action value ${params.actionValue} is ${zScore.toFixed(1)}σ from mean (${params.historicalMeanValue})`,
    };
  }

  if (params.recentExecutionCount > params.burstThreshold) {
    return {
      anomalyDetected: true,
      description: `execution burst: ${params.recentExecutionCount} in window (threshold: ${params.burstThreshold})`,
    };
  }

  if (params.isOffHours && params.actionValue > params.historicalMeanValue * 2) {
    return {
      anomalyDetected: true,
      description: `off-hours execution with above-average value ${params.actionValue}`,
    };
  }

  return { anomalyDetected: false, description: null };
}

File: python/atlasent/governance/autonomous_financial.py
"""Autonomous Financial Execution — priority 5.

Mirrors ``atlasent-sdk/typescript/src/autonomousFinancial.ts``.

Key divergences from the legacy ``atlasent.governance.autonomous_financial``
in the orchestration repo (which this canonical port supersedes):

- Allowlist-based ``permitted_action_types`` (not denylist).
- Per-action-type ``ExecutionCeiling`` with optional ``max_daily_count``.
- ``max_risk_tier`` ceiling with strict ordinal comparison
  (low < medium < high < critical).
- Bounds expiry via ``expires_at`` (ISO-8601 string, lex-comparable).
- ``detect_autonomous_anomaly`` covers z-score (3σ), burst, and
  off-hours-with-above-mean. The legacy module had only z-score.
- The legacy ``max_counterparty_exposure`` field is intentionally NOT
  ported; it is not in the canonical TS module or in migration
  ``autonomous_execution_bounds`` and would be enforcing a constraint the
  rest of the system doesn't know about.

Wire-stable as ``autonomous_financial.v1``.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional, Sequence

from .financial_action import CurrencyCode, FinancialActionType, FinancialRiskTier

_RISK_TIER_ORDER: dict[str, int] = {
    "low": 1,
    "medium": 2,
    "high": 3,
    "critical": 4,
}


@dataclass(frozen=True)
class ExecutionCeiling:
    """Per-action-type execution ceiling."""

    action_type: FinancialActionType
    per_execution_max: float
    currency: CurrencyCode
    max_daily_count: Optional[int]
    require_permit: bool = False


@dataclass(frozen=True)
class AutonomousExecutionBounds:
    """Authority bounds for an autonomous financial agent.

    Stored in migration table ``autonomous_execution_bounds``. Agents may
    not self-modify their own bounds.
    """

    bounds_id: str
    org_id: str
    agent_id: str
    agent_name: str
    permitted_action_types: Sequence[FinancialActionType]
    ceilings: Sequence[ExecutionCeiling]
    daily_aggregate_ceiling: float
    aggregate_currency: CurrencyCode
    max_risk_tier: FinancialRiskTier
    require_runtime_verification: bool
    anomaly_detection_enabled: bool
    created_at: str
    expires_at: Optional[str]
    active: bool


@dataclass(frozen=True)
class AutonomousExecutionRecord:
    """Record of an autonomous execution attempt (audit trail)."""

    record_id: str
    agent_id: str
    org_id: str
    action_type: FinancialActionType
    action_value: float
    currency: CurrencyCode
    permitted: bool
    denial_reason: Optional[str]
    permit_id: Optional[str]
    anomaly_detected: bool
    anomaly_description: Optional[str]
    attempted_at: str
    executed_at: Optional[str]


@dataclass(frozen=True)
class AutonomousExecutionCheckResult:
    """Result of checking whether an autonomous execution is within bounds."""

    permitted: bool
    action_type_permitted: bool
    within_execution_ceiling: bool
    within_daily_aggregate: bool
    within_risk_tier: bool
    bounds_active: bool
    bounds_not_expired: bool
    applicable_ceiling: Optional[ExecutionCeiling]
    denial_reason: Optional[str]
    violations: Sequence[str]


@dataclass(frozen=True)
class AnomalyDetectionResult:
    """Result of ``detect_autonomous_anomaly``."""

    anomaly_detected: bool
    description: Optional[str]


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def check_autonomous_bounds(
    *,
    bounds: AutonomousExecutionBounds,
    action_type: FinancialActionType,
    action_value: float,
    currency: CurrencyCode,
    risk_tier: FinancialRiskTier,
    current_daily_aggregate: float,
    current_daily_count: dict[str, int],
    now: Optional[str] = None,
) -> AutonomousExecutionCheckResult:
    """Check whether an autonomous execution is within declared bounds.

    Mirrors ``checkAutonomousBounds`` in TS step-for-step.
    """
    violations: list[str] = []
    now_iso = now if now is not None else _now_iso()

    bounds_active = bounds.active
    if not bounds_active:
        violations.append("agent execution bounds are inactive")

    bounds_not_expired = bounds.expires_at is None or bounds.expires_at > now_iso
    if not bounds_not_expired:
        violations.append(f"agent bounds expired at {bounds.expires_at}")

    action_type_permitted = action_type in bounds.permitted_action_types
    if not action_type_permitted:
        violations.append(
            f"action type {action_type} not in agent's permitted set"
        )

    applicable_ceiling: Optional[ExecutionCeiling] = next(
        (c for c in bounds.ceilings if c.action_type == action_type), None
    )

    within_execution_ceiling = True
    if applicable_ceiling is not None:
        if action_value > applicable_ceiling.per_execution_max:
            within_execution_ceiling = False
            violations.append(
                f"value {action_value} exceeds per-execution ceiling "
                f"{applicable_ceiling.per_execution_max} {applicable_ceiling.currency}"
            )
        if applicable_ceiling.max_daily_count is not None:
            today_count = current_daily_count.get(action_type, 0)
            if today_count >= applicable_ceiling.max_daily_count:
                within_execution_ceiling = False
                violations.append(
                    f"daily count {today_count} at or exceeds limit "
                    f"{applicable_ceiling.max_daily_count} for {action_type}"
                )

    within_daily_aggregate = (
        current_daily_aggregate + action_value <= bounds.daily_aggregate_ceiling
    )
    if not within_daily_aggregate:
        violations.append(
            f"daily aggregate {current_daily_aggregate + action_value} would exceed "
            f"ceiling {bounds.daily_aggregate_ceiling} {bounds.aggregate_currency}"
        )

    within_risk_tier = (
        _RISK_TIER_ORDER[risk_tier] <= _RISK_TIER_ORDER[bounds.max_risk_tier]
    )
    if not within_risk_tier:
        violations.append(
            f"action risk tier {risk_tier} exceeds agent max {bounds.max_risk_tier}"
        )

    permitted = (
        bounds_active
        and bounds_not_expired
        and action_type_permitted
        and within_execution_ceiling
        and within_daily_aggregate
        and within_risk_tier
    )

    return AutonomousExecutionCheckResult(
        permitted=permitted,
        action_type_permitted=action_type_permitted,
        within_execution_ceiling=within_execution_ceiling,
        within_daily_aggregate=within_daily_aggregate,
        within_risk_tier=within_risk_tier,
        bounds_active=bounds_active,
        bounds_not_expired=bounds_not_expired,
        applicable_ceiling=applicable_ceiling,
        denial_reason=(
            None if permitted else (violations[0] if violations else "execution out of bounds")
        ),
        violations=tuple(violations),
    )


def detect_autonomous_anomaly(
    *,
    action_value: float,
    historical_mean_value: float,
    historical_std_dev: float,
    recent_execution_count: int,
    burst_threshold: int,
    is_off_hours: bool,
) -> AnomalyDetectionResult:
    """Detect a potential anomaly in autonomous execution.

    Returns an ``AnomalyDetectionResult`` with a description when an anomaly
    is detected, or ``description=None``. Mirrors ``detectAutonomousAnomaly``
    in TS.
    """
    z_score = (
        abs(action_value - historical_mean_value) / historical_std_dev
        if historical_std_dev > 0
        else 0.0
    )

    if z_score > 3:
        return AnomalyDetectionResult(
            anomaly_detected=True,
            description=(
                f"action value {action_value} is {z_score:.1f}σ from mean "
                f"({historical_mean_value})"
            ),
        )

    if recent_execution_count > burst_threshold:
        return AnomalyDetectionResult(
            anomaly_detected=True,
            description=(
                f"execution burst: {recent_execution_count} in window "
                f"(threshold: {burst_threshold})"
            ),
        )

    if is_off_hours and action_value > historical_mean_value * 2:
        return AnomalyDetectionResult(
            anomaly_detected=True,
            description=f"off-hours execution with above-average value {action_value}",
        )

    return AnomalyDetectionResult(anomaly_detected=False, description=None)


__all__ = [
    "AnomalyDetectionResult",
    "AutonomousExecutionBounds",
    "AutonomousExecutionCheckResult",
    "AutonomousExecutionRecord",
    "ExecutionCeiling",
    "check_autonomous_bounds",
    "detect_autonomous_anomaly",
]
